### PR TITLE
update sagemaker tests for tensorflow 2

### PR DESCRIPTION
**Pull request by hsl89**
*GitHub Issue #, if available:*

### Description
This PR tackles the SM local test and SM test failures in the mainline pipeline. It also conducts other refactoring to make the tests more robust. Additionally, it takes care of the fact that the pushed changes do not break the tests for the older TF versions.

### Tests run
Ran the tests in the SM tests and SM local test in the CI.

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true` or `neuron_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/local/test_training.py [hunk 1]
```diff
@@ -72,7 +74,11 @@ def test_mnist_cpu(sagemaker_local_session, docker_image, tmpdir, framework_vers
                     output_path=output_path,
                     training_data_path='file://{}'.format(
                         os.path.join(RESOURCE_PATH, 'mnist', 'data')))
-    _assert_files_exist_in_tar(output_path, ['my_model.h5'])
+    # _assert_checkpoint_exists_v2(output_path)
+    ############################################################################
+    #######TODO: This is something that needs to be confirmed from TF2.6 #######
```
Comment (shantanutrip) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/local/test_training.py:79]: I have made this fix in which I am not using `_assert_files_exist_in_tar` function instead of  `_assert_checkpoint_exists_v2`. I want the change to be verified.

I am not sure why we were using [_assert_checkpoint_exists_v2](https://github.com/aws/deep-learning-containers/blob/2b199676aee4d5e52af4f53e2c7631095b20b13a/test/sagemaker_tests/tensorflow/tensorflow2_training/integration/local/test_training.py#L179) to search for checkpoint in the s3 bucket? I ask this because, the [`output_path`](https://github.com/aws/deep-learning-containers/blob/2b199676aee4d5e52af4f53e2c7631095b20b13a/test/sagemaker_tests/tensorflow/tensorflow2_training/integration/local/test_training.py#L67) variable looks like `file://{}`. As per the sagemaker estimator [documentation](https://sagemaker.readthedocs.io/en/stable/api/training/estimators.html), `output_path` with `file://{}` format saves the training results in the local folder. That being said, on investigation, I found out that 2 tar files, namely `model.tar.gz` and `output.tar.gz` are being saved successfully in the desired location. Once I untar `model.tar.gz`, I get a bunch of files and folders. One of the untarred files turns out to be a pickle file called `saved_model.pb`.

This might pass the test, but I am not sure if this is what the team desired. I would like the TF team to look into the changes I made in this commit and verify if the things are working as intended.

Comment (hsl89) [reply test/sagemaker_tests/tensorflow/tensorflow2_training/integration/local/test_training.py:79]: It's because `_assert_checkpoints_exitsts_v2` is agnostic about the naming of the files in `output_path`, tensorflow might save the checkpoint with different conventions across different versions.

Comment (shantanutrip) [reply test/sagemaker_tests/tensorflow/tensorflow2_training/integration/local/test_training.py:79]: Do we need to check the file in s3 bucket since we are saving the output locally?
We can just change [this](https://github.com/aws/deep-learning-containers/blob/2b199676aee4d5e52af4f53e2c7631095b20b13a/test/sagemaker_tests/tensorflow/tensorflow2_training/integration/local/test_training.py#L80) line to _assert_files_exist_in_tar(output_path, [])  such that it becomes agnostic in nature, right?

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py [hunk 2]
```diff
@@ -69,7 +70,8 @@ def test_distributed_mnist_no_ps(sagemaker_session, ecr_image, instance_type, fr
         path=os.path.join(resource_path, 'mnist', 'data'),
         key_prefix='scriptmode/mnist')
     estimator.fit(inputs, job_name=unique_name_from_base('test-tf-sm-distributed-mnist'))
-    _assert_s3_file_exists(sagemaker_session.boto_region_name, estimator.model_data)
+    _assert_checkpoint_exists_v2(estimator.model_dir)
+    # _assert_s3_file_exists(sagemaker_session.boto_region_name, estimator.model_data)
```
Comment (SergTogul) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:74]: Why don't we delete it? What is a purpose of __assert_s3_file_exists_

**Review by SergTogul — APPROVED** [review]

**Comment by tejaschumbalkar** [comment]
Please update the PR description

**Comment by shantanutrip** [comment]
> Please update the PR description

Done

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py [hunk 3]
```diff
@@ -121,11 +153,16 @@ def test_s3_plugin(sagemaker_session, ecr_image, instance_type, region, framewor
...
-    estimator.fit('s3://sagemaker-sample-data-{}/tensorflow/mnist'.format(region),
-                  job_name=unique_name_from_base('test-tf-sm-s3-mnist'))
-    _assert_s3_file_exists(region, estimator.model_data)
-    _assert_checkpoint_exists(region, estimator.model_dir, 200)
 
+    inputs = estimator.sagemaker_session.upload_data(
+        path=os.path.join(resource_path, 'mnist', 'data-distributed'),
+        key_prefix='scriptmode/mnist-distributed')
+    estimator.fit(inputs,job_name=unique_name_from_base('test-tf-sm-s3-mnist'))
+    print("=========== Model data location ===============")
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:161]: nit: lets remove this

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py [hunk 4]
```diff
@@ -219,20 +256,19 @@ def test_smdataparallel_smmodelparallel_mnist(sagemaker_session, instance_type,
...
-        Bucket=bucket, Prefix=prefix
-    )['Contents']
-    assert len(ckpt_content) > 0, "checkpoint directory is emptry"
-    print(ckpt_content)
+    _assert_s3_file_exists(region, os.path.join(model_dir, 'checkpoint'))
+    _assert_s3_file_exists(region,
+                           os.path.join(model_dir, 'model.ckpt-{}.index'.format(checkpoint_number)))
+
 
 def _assert_checkpoint_exists(region, model_dir, checkpoint_number):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:268]: nit: Should we rename it to `_assert_checkpoint_exists_v1` for better clarity?

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py [hunk 5]
```diff
@@ -26,6 +27,15 @@
...
 from .timeout import timeout
 
+def can_run_mnist_estimator(ecr_image):
+    _, image_framework_version = get_framework_and_version_from_tag(ecr_image)
+    return Version(image_framework_version) in SpecifierSet("<2.6")
+
+
+def validate_or_skip_test(ecr_image):
+    if not can_run_mnist_estimator(ecr_image):
+        pytest.skip("Mnist Estimator related scripts can only run for versions < 2.6")
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:37]: can we provide more information on why this is the case?

**Comment by hsl89** [comment]
LGTM

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py [hunk 6]
```diff
@@ -219,20 +256,19 @@ def test_smdataparallel_smmodelparallel_mnist(sagemaker_session, instance_type,
     estimator.fit()
 
 
-def _assert_checkpoint_exists_v2(s3_model_dir):
+def _assert_checkpoint_exists_v2(region, model_dir, checkpoint_number):
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:259]: nit: can you clearly describe what you mean by v2 style checkpoints? Assuming this doesn't mean tensorflow2, since all of these tests run on TF2

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py [hunk 7]
```diff
@@ -76,30 +86,52 @@ def test_distributed_mnist_no_ps(sagemaker_session, ecr_image, instance_type, fr
...
+    inputs = estimator.sagemaker_session.upload_data(
+        path=os.path.join(resource_path, 'mnist', 'data-distributed'),
+        key_prefix='scriptmode/mnist-distributed')
+    estimator.fit(inputs, job_name=unique_name_from_base('test-tf-sm-distributed-mnist'))
+    _assert_checkpoint_exists(sagemaker_session.boto_region_name, estimator.model_dir, 200)
+
+@pytest.mark.model("mnist")
+@pytest.mark.multinode(2)
+@pytest.mark.integration("parameter server")
+def test_distributed_mnist_custom_ps(sagemaker_session, ecr_image, instance_type, framework_version):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:110]: Is this test compatible with all versions of TF DLC?

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py (hunk 3 above)
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:161]: nit: we should configure a logger using python logging library so that we can get things like timestamps and other debug information. This can be done as a followup, since I see we're using `print` in other places in the code, and this is consistent with the existing setup.

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py (hunk 7 above)
Comment (shantanutrip) [reply test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:110]: Yes, as said by the TF team, mnist_custom script is compatible for all tf2 versions that we support.

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py [hunk 8]
```diff
@@ -76,30 +86,52 @@ def test_distributed_mnist_no_ps(sagemaker_session, ecr_image, instance_type, fr
...
+                           instance_type=instance_type,
+                           sagemaker_session=sagemaker_session,
+                           image_uri=ecr_image,
+                           framework_version=framework_version)
+    inputs = estimator.sagemaker_session.upload_data(
+        path=os.path.join(resource_path, 'mnist', 'data-distributed'),
+        key_prefix='scriptmode/mnist-distributed')
+    estimator.fit(inputs, job_name=unique_name_from_base('test-tf-sm-distributed-mnist'))
+    _assert_checkpoint_exists(sagemaker_session.boto_region_name, estimator.model_dir, 200)
+
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_mnist.py:106]: nit: we use black code formatter (https://pypi.org/project/black/) - generally there should be two new lines between functions. You can execute it with

```
pip install black
black -l 120 /path/to/file.py
```

File: test/sagemaker_tests/tensorflow/tensorflow2_training/resources/mnist/mnist_custom.py [hunk 9]
```diff
@@ -15,9 +15,17 @@
 import os
 import json
 import tensorflow as tf
-import tensorflow_io as tfio
+
+tf_major, tf_minor, _ = tf.__version__.split('.')
+if int(tf_major) > 2 or (int(tf_major) == 2 and int(tf_minor) >= 6):
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/resources/mnist/mnist_custom.py:20]: I believe `packaging` should already be installed - you could do something like this, which might be a little more straightforward and clear

```
from packaging.version import Version

if Version(tf.__version__) >= Version("2.6.0"):
    import tensorflow_io as tfio
```

**Comment by arjkesh** [comment]
@shantanutrip will address comments in subsequent PR - leaving approval to unblock testing

**Review by arjkesh — APPROVED** [review]
